File: golang/3531.count-covered-buildings.go

import (
	"sort"
)

func countCoveredBuildings(n int, buildings [][]int) int {
	covered := 0
	hor := make(map[int][]int)
	ver := make(map[int][]int)

	for _, b := range buildings {
		x, y := b[0], b[1]

		hor[x] = append(hor[x], y)
		ver[y] = append(ver[y], x)
	}

	for _, ys := range hor {
		sort.Ints(ys)
	}

	for _, xs := range ver {
		sort.Ints(xs)
	}

	for x, ys := range hor {
		for i := 1; i < len(ys)-1; i++ {
			y := ys[i]

			if ver[y][0] < x && x < ver[y][len(ver[y])-1] {
				covered++
			}
		}
	}

	return covered
}

File: cpp/3531.count-covered-buildings.cpp
#include <unordered_map>
#include <vector>
using std::vector, std::unordered_map;

class Solution {
public:
  int countCoveredBuildings(int n, vector<vector<int>> &buildings) {
    int covered = 0;
    unordered_map<int, vector<int>> hor;
    unordered_map<int, vector<int>> ver;

    for (auto &b : buildings) {
      int x = b[0], y = b[1];

      hor[x].push_back(y);
      ver[y].push_back(x);
    }

    for (auto &[_, ys] : hor) {
      std::sort(ys.begin(), ys.end());
    }

    for (auto &[_, xs] : ver) {
      std::sort(xs.begin(), xs.end());
    }

    for (auto &[x, ys] : hor) {

      for (int i = 1; i < ys.size() - 1; i++) {

        int y = ys[i];

        if (ver[y].front() < x && x < ver[y].back()) {
          covered++;
        }
      }
    }

    return covered;
  }
};